prow/git/v2/client_factory.go

```diff
@@ -339,21 +346,69 @@ func (c *clientFactory) ClientForWithRepoOpts(org, repo string, repoOpts RepoOpt
 		// something unexpected happened
 		return nil, err
 	} else {
-		// we have cloned the repo previously, but will refresh it
-		if err := cacheClientCacher.RemoteUpdate(); err != nil {
-			return nil, err
+		// We have cloned the repo previously, but will refresh it. By default
+		// we refresh all refs with a call to `git remote update`.
+		//
+		// This is the default behavior if FetchCommits is empty or nil (i.e.,
+		// when we don't define a targeted list of commits to fetch directly).
+		if len(repoOpts.FetchCommits) == 0 {
+			if err := cacheClientCacher.RemoteUpdate(); err != nil {
+				return nil, err
+			}
 		}
 	}
 
-	// initialize the new derivative repo from the cache
+	// Clone the secondary clone (local clone with checked out files) from the
+	// primary (bare repo).
 	if err := repoClientCloner.CloneWithRepoOpts(cacheDir, repoOpts); err != nil {
 		return nil, err
 	}
 
+	// Here we do a targeted fetch, if the repoOpts for the secondary clone
+	// asked for it.
+	//
+	// TODO(listx): Change ensureCommits() so that it takes a "cacher" interface
+	// instead of a generic RepoClient. However, if ShareObjectsWithSourceRepo
+	// was not set, then we cannot run ensureCommits() with the cacher because
+	// the cacher will run on the primary, not secondary (and without --shared,
+	// this will have no effect on the secondary); we have to handle that case
+	// gracefully (perhaps with some logging also).
+	if len(repoOpts.FetchCommits) > 0 {
+		// Targeted fetch. Only fetch those commits which we want, and only
+		// if they are missing.
+		if err := ensureCommits(repoClient, repoOpts); err != nil {
+			return nil, err
+		}
+	}
+
 	return repoClient, nil
 }
 
 // Clean removes the caches used to generate clients
 func (c *clientFactory) Clean() error {
 	return os.RemoveAll(c.cacheDir)
 }
+
+func ensureCommits(repoClient RepoClient, repoOpts RepoOpts) error {
+	fetchArgs := []string{}
+
+	if repoOpts.NoFetchTags {
+		fetchArgs = append(fetchArgs, "--no-tags")
+	}
+
+	// For each commit SHA, check if it already exists. If so, don't bother
+	// fetching it.
+	for _, commitSHA := range repoOpts.FetchCommits {
+		if exists, _ := repoClient.ObjectExists(commitSHA); exists {
+			continue
+		}
+
+		fetchArgs = append(fetchArgs, commitSHA)
+	}
+
+	if err := repoClient.Fetch(fetchArgs...); err != nil {
+		return fmt.Errorf("failed to fetch %s: %v", fetchArgs, err)
+	}
```

Comment:
Is this only fetching between the primary and secondary clones? We need to fetch from upstream to the primary clone then fetch to the secondary clone, right?
Or is it the case that this fetch for the secondary clone actually reaches the upstream because the secondary is a `--shared` repo?

Reply:
> Is this only fetching between the primary and secondary clones? We need to fetch from upstream to the primary clone then fetch to the secondary clone, right?

From our internal logs for the 1st attempt, the fetch will happen but with an explicit `<remote>` argument, where `<remote>` looks like "https://...". So even though the repoClient is the secondary clone, it will construct a fetch command that looks like "git fetch https://... ...".

> Or is it the case that this fetch for the secondary clone actually reaches the upstream because the secondary is a --shared repo?

Without the additional `<remote>` argument, it would only try to fetch from the primary, and the fetch will fail even with --shared.

So in summary it looks like the above code should be OK.

Reply:
Gonna rebase this on top of https://github.com/kubernetes/test-infra/pull/30488 to test the "fetch, don't clone from scratch" code path.

Reply:
Done.